### ISSUE TITLE
Off-chip memory used for very small models

### ISSUE THREAD

**Issue by marcin-kamionowski** [google-coral/edgetpu#324]
Hi!
We observe quite often cases when small amount off-chip memory is used.
Not sure why. It says that 7.77MiB is available so everything should be placed on chip:
> $ edgetpu_compiler -s tl_obj_experimental.tflite 
> Edge TPU Compiler version 15.0.340273435
> 
> Model compiled successfully in 128 ms.
> 
> Input model: tl_obj_experimental.tflite
> Input size: 76.14KiB
> Output model: tl_obj_experimental_edgetpu.tflite
> Output size: 152.54KiB
> On-chip memory used for caching model parameters: 85.75KiB
> On-chip memory remaining for caching model parameters: 7.77MiB
> Off-chip memory used for streaming uncached model parameters: 5.50KiB
> Number of Edge TPU subgraphs: 1
> Total number of operations: 15
> Operation log: tl_obj_experimental_edgetpu.log
> 
> Operator                       Count      Status
> 
> SOFTMAX                        1          Mapped to Edge TPU
> FULLY_CONNECTED                2          Mapped to Edge TPU
> ADD                            4          Mapped to Edge TPU
> MUL                            4          Mapped to Edge TPU
> CONV_2D                        4          Mapped to Edge TPU

This is the model: [tl_obj_experimental.zip](https://github.com/google-coral/edgetpu/files/6048517/tl_obj_experimental.zip)

**Comment by Namburger** [comment]
@Naveen-Dodda could you take a look at this?

**Comment by Namburger** [comment]
@marcin-kamionowski 
We use offchip  memory to stream the output tensors so there are some amount of offchip memory expected.
to test this, you try changing the output tensor size and check the compiler output again

**Comment by marcin-kamionowski** [comment]
This is the model in Netron:
![Screenshot from 2021-03-01 09-56-10](https://user-images.githubusercontent.com/32487822/109474316-af410100-7a74-11eb-842c-4d3b02fb410a.png)
Output tensor is 1x4 so it is very small. Are you referring to this size?
I also checked inner operations sizes - biggest one is 22x10x32. I compared this to other model that do not use off-chip memory (144x144x32) so I'm confused what is causing this issue.

**Comment by ricardodeazambuja** [comment]
I'm having a problem a lot like the one above. My model had all ops mapped to the Edge TPU, but the size grew from 1.4MiB (original tflite) to around 40MiB and off-chip memory used is almost 40MiB! Another different model that was 3MiB only grew to 3.7MiB and had zero off-chip memory used.

I tested the problematic model with (original tflite) and without the Edge TPU. The pure tflite is less than 2x slower. My model also makes use of 2 fully connected layers.

**Comment by Naveen-Dodda** [comment]
Hello, 

Size of the model growing from 1.4MiB to 40MiB, is abnormal. Could you please share your model here, and share details of your training/tflite_convert environment.

**Comment by ricardodeazambuja** [comment]
Hi @Naveen-Dodda,

Thank you for your feedback. The model is the tiny LPRNet from this [repo](https://github.com/GreenWaves-Technologies/licence_plate_recognition).
Here is the tflite file: [tflite_model.zip](https://github.com/google-coral/edgetpu/files/6450295/tflite_model.zip).

Output from edgetpu_compiler:
```
Edge TPU Compiler version 15.0.340273435

Model compiled successfully in 1181 ms.

Input model: model.tflite
Input size: 1.40MiB
Output model: model_edgetpu.tflite
Output size: 39.72MiB
On-chip memory used for caching model parameters: 1.68MiB
On-chip memory remaining for caching model parameters: 5.99MiB
Off-chip memory used for streaming uncached model parameters: 37.92MiB
Number of Edge TPU subgraphs: 1
Total number of operations: 25
Operation log: model_edgetpu.log
See the operation log file for individual operation details.
```
And the log:
```
Edge TPU Compiler version 15.0.340273435
Input: model.tflite
Output: model_edgetpu.tflite

Operator                       Count      Status

CONV_2D                        16         Mapped to Edge TPU
LOGISTIC                       1          Mapped to Edge TPU
FULLY_CONNECTED                2          Mapped to Edge TPU
RESHAPE                        2          Mapped to Edge TPU
CONCATENATION                  1          Mapped to Edge TPU
MAX_POOL_2D                    3          Mapped to Edge TPU

```

**Comment by ricardodeazambuja** [comment]
```
input_shape = (1,88,71)
inputs = keras.Input(shape=input_shape)
fl1 = layers.Flatten()(inputs)
d1 = layers.Dense(64, activation="relu")(fl1)
d2 = layers.Dense(88*71, activation="linear")(d1)
r1 = layers.Reshape((1,88,71))(d2)
ct1 = layers.Concatenate()([inputs, r1])
c15 = layers.Conv2D(71,1,padding='same',activation=None)(ct1)
outputs = layers.Reshape((88,71))(c15)
```
The model above, when compiled by edgetpu_compiler, has the same bug I mentioned before.

However, this model compiles just fine:
```
input_shape = (6248)
inputs = keras.Input(shape=input_shape)
d1 = layers.Dense(64, activation="relu")(inputs)
d2 = layers.Dense(88*71, activation="linear")(d1)
r1 = layers.Reshape((1,88,71))(d2)
r2 = layers.Reshape((1,88,71))(inputs)
ct1 = layers.Concatenate()([r2, r1])
c15 = layers.Conv2D(71,1,padding='same',activation=None)(ct1)
outputs = layers.Reshape((88,71))(c15)
```

I think the edgetpu_compiler has problems with the flatten layer.

**Comment by ricardodeazambuja** [comment]
I just discovered something quite funny. If I move the output of the model, without changing the graph (so I still can access the final value), the edgetpu_compiler stops crazily increasing the size of the model. Same model, just the definition of which tensor is the output changed :D

**Comment by ricardodeazambuja** [comment]
My hope didn't last, only the original tflite model keeps the whole graph, the compiled one is chopped. However, when I use this:
```edgetpu_compiler -i Conv_14/Relu,Squeeze -s -a model.tflite```
The output is not what I would expect as it maps one FULLY_CONNECTED op but not the other, probably "-a" magic:
```
Edge TPU Compiler version 15.0.340273435

Model compiled successfully in 273 ms.

Input model: model.tflite
Input size: 1.40MiB
Output model: model_edgetpu.tflite
Output size: 1.84MiB
On-chip memory used for caching model parameters: 1.30MiB
On-chip memory remaining for caching model parameters: 6.89MiB
Off-chip memory used for streaming uncached model parameters: 0.00B
Number of Edge TPU subgraphs: 3
Total number of operations: 25
Operation log: model_edgetpu.log

Model successfully compiled but not all operations are supported by the Edge TPU. A percentage of the model will instead run on the CPU, which is slower. If possible, consider updating your model to use only operations supported by the Edge TPU. For details, visit g.co/coral/model-reqs.
Number of operations that will run on Edge TPU: 23
Number of operations that will run on CPU: 2

Operator                       Count      Status

RESHAPE                        2          Mapped to Edge TPU
MAX_POOL_2D                    3          Mapped to Edge TPU
CONCATENATION                  1          More than one subgraph is not supported
FULLY_CONNECTED                1          More than one subgraph is not supported
FULLY_CONNECTED                1          Mapped to Edge TPU
LOGISTIC                       1          Mapped to Edge TPU
CONV_2D                        16         Mapped to Edge TPU
```
This is the compiled model:
![image](https://user-images.githubusercontent.com/6606382/119703997-0a557a80-be4f-11eb-8588-cdd9638b1992.png)

And the original one (it was too big, so here is just the ops where the graph splits):
![image](https://user-images.githubusercontent.com/6606382/119704307-68825d80-be4f-11eb-9306-068754f2eb00.png)

**Comment by Naveen-Dodda** [comment]
@ricardodeazambuja In the above comments you mentioned that removing  flatten layer fixes you issues of large export size. Could you please help me narrow down the issue you are having. 

Thanks

**Comment by ricardodeazambuja** [comment]
@Naveen-Dodda Yes, sure.
My last experiment was splitting the model into two, exactly before the first fully connected layer. Therefore I had two models, one from the original input until the Conv2D where the model branches, and another where that Conv2D was replaced with an input. The compiler creates normal sized models. If I add a reshape operation after the Conv2D or before the first fully connected, the model size increases crazily.

**Comment by Naveen-Dodda** [comment]
Okay, got it but the other reshape op present in 2nd graph doesn't create this problem.

**Comment by ricardodeazambuja** [comment]
What is the black magic behind "Off-chip memory used for streaming uncached model parameter"?
```
Edge TPU Compiler version 15.0.340273435

Model compiled successfully in 1379 ms.

Input model: model.tflite
Input size: 4.41MiB
Output model: model_edgetpu.tflite
Output size: 5.49MiB
On-chip memory used for caching model parameters: 1.92MiB
On-chip memory remaining for caching model parameters: 1.75KiB
Off-chip memory used for streaming uncached model parameters: 3.23MiB
Number of Edge TPU subgraphs: 1
Total number of operations: 136
Operation log: model_edgetpu.log

Model successfully compiled but not all operations are supported by the Edge TPU. A percentage of the model will instead run on the CPU, which is slower. If possible, consider updating your model to use only operations supported by the Edge TPU. For details, visit g.co/coral/model-reqs.
Number of operations that will run on Edge TPU: 135
Number of operations that will run on CPU: 1

Operator                       Count      Status

RESHAPE                        15         Mapped to Edge TPU
CONCATENATION                  3          Mapped to Edge TPU
MAX_POOL_2D                    3          Mapped to Edge TPU
ADD                            10         Mapped to Edge TPU
CUSTOM                         1          Operation is working on an unsupported data type
CONV_2D                        73         Mapped to Edge TPU
DEPTHWISE_CONV_2D              29         Mapped to Edge TPU
LOGISTIC                       2          Mapped to Edge TPU
```
Another curiosity. This model is made of two other tflite models in parallel (two inputs). However, if I compile the models for shared cache (using the original two tflite files), depending on the order it changes the amount of "Off-chip memory used for streaming uncached model parameters" assigned to each model.

**P.S.: when I compile the models individually, this is the result for any of the models:**
```
Off-chip memory used for streaming uncached model parameters: 0.00B
```

**Comment by hjonnala** [comment]
@marcin-kamionowski can you please share your saved model and quantization script?
